File: 常用算法模板/并查集/模板.go
package main
import "fmt"

/*
并查集，判连通用
*/
type UnionFindSet struct {
	father  []int // 存储结点的父亲
	height  []int // 存储结点的父亲
	nodeNum int   // 总结点个数
}

func (us *UnionFindSet) InitUnionSet(n int) {
	us.nodeNum = n+1 // 不加也可以，有人喜欢以0开头
	us.father = make([]int, us.nodeNum)
	us.height = make([]int, us.nodeNum)
	for i, _ := range us.father {
		us.father[i] = i
		us.height[i] = 1
	}
}

// 查询父结点
func (us *UnionFindSet) Find(x int) int {
	for us.father[x] != x {
		x = us.father[x]
	}
	return x
}

//合并结点
func (us *UnionFindSet) Union(x, y int) bool {
	x = us.Find(x)
	y = us.Find(y)
	if x == y {
		return false
	}
	us.father[x] = y
	return true
}

func (us *UnionFindSet) FindV2(x int) int {
	root := x // 保存好路径上的头结点
	for us.father[root] != root {
		root = us.father[root]
	}
	/*
	从头结点开始一直往根上遍历
	把所有结点的father直接指向root。
	*/
	for us.father[x] != x {
		// 一定要先保存好下一个结点，下一步是要对us.father[x]进行赋值
		temp := us.father[x]
		us.father[x] = root
		x = temp
	}

	return root
}

/*
需要加入height[]属性，初始化为1.
*/
//合并结点
func (us *UnionFindSet) UnionV2(x, y int) bool {
	x = us.Find(x)
	y = us.Find(y)
	if x == y {
		return false
	}
	if us.height[x]<us.height[y] {
		us.father[x]=y
	} else if us.height[x]>us.height[y] {
		us.father[y]=x
	} else {
		us.father[x] = y
		us.height[y]++
	}
	return true
}


func findCircleNum(M [][]int) int {
	us := &UnionFindSet{}
	us.InitUnionSet(len(M))
	for i,row:=range M{
		for j,isFriend :=range row {
			if isFriend==1 {us.UnionV2(i,j)} // 是朋友则合并
		}
	}

	friendCircle :=0
	for i:=0;i<len(M);i++ {
		if us.FindV2(i)==i { friendCircle++} // 查看集合代表个数
	}

	return friendCircle
}
/*
[[1,1,0],[1,1,0],[0,0,1]]
[[1]]
[[1,0,0],[0,1,0],[0,0,1]]
[[1,0,1],[0,1,0],[1,0,1]]
[[1,0,1,0],[0,1,0,0],[1,0,1,1],[0,0,1,1]]
*/

func main() {
	us:= &UnionFindSet{}
	us.InitUnionSet(11)
	us.UnionV2(1, 2)
	us.UnionV2(10, 9)
	us.UnionV2(7,8)
	us.UnionV2(8,3)
	us.UnionV2(9,8)

	for i:=0;i<=10; i++  {
		fmt.Println(i, us.FindV2(i))
	}
}

File: 常用算法模板/并查集/模板.cpp

#include <iostream>
#include<vector>
#include<string>
#include<cstring>
#include<stdio.h>
#include <queue>
#include<map>
#include <algorithm>

using namespace std;


class UnionFindSet {
private:
    vector<int> father; // 父结点定义，father[i]=i时，i为本集合的代表
    vector<int> height; // 代表树高度，初始为1
    int nodeNum; // 集合中的点数

public:
    UnionFindSet(int n); // 初始化
    bool Union(int x, int y); // 合并
    int Find(int x);

    bool UnionV2(int x, int y); // 合并
    int FindV2(int x);
};

UnionFindSet::UnionFindSet(int n) : nodeNum(n + 1) {
    father = vector<int>(nodeNum);
    height = vector<int>(nodeNum);
    for (int i = 0; i < nodeNum; ++i) father[i] = i, height[i] = 1; // 初始为自己
}

int UnionFindSet::Find(int x) {
    while (father[x] != x) x = father[x];
    return x;
}

bool UnionFindSet::Union(int x, int y) {
    x = Find(x);
    y = Find(y);

    if (x == y)return false;
    father[x] = y;
    return true;
}


int UnionFindSet::FindV2(int x) {
    int root = x; // 保存好路径上的头结点
    while (father[root] != root) {
        root = father[root];
    }
/*
从头结点开始一直往根上遍历
把所有结点的father直接指向root。
*/
    while(father[x] != x) {
        // 一定要先保存好下一个结点，下一步是要对father[x]进行赋值
        int temp = father[x];
        father[x] = root;
        x = temp;
    }

    return root;
}

/*
需要加入height[]属性，初始化为1.
*/
//合并结点
bool UnionFindSet::UnionV2(int x, int y) {
    x = Find(x);
    y = Find(y);
    if (x == y) {
        return false;
    }
    if (height[x] < height[y]) {
        father[x] = y;
    } else if (height[x] > height[y]) {
        father[y] = x;
    } else {
        father[x] = y;
        height[y]++;
    }
    return true;
}

class IndexMaker {
private:
    int curUsedIndex;
    map<string, int> indMap;
public:
    IndexMaker();

    int GetIndByName(string name);
};

IndexMaker::IndexMaker() : curUsedIndex(0) {
    indMap.clear();
}

int IndexMaker::GetIndByName(string name) {
    if (indMap.count(name) == 0) indMap[name] = curUsedIndex++;
    return indMap[name];
}

int main() {

    int n;
    cin >> n;
    UnionFindSet us(2 * n); // 最多有可能2n个人

    string a, b;
    int o;
    IndexMaker indMaker;

    for (int i = 0; i < n; ++i) {
        cin >> o >> a >> b;
        int indexa = indMaker.GetIndByName(a), indexb = indMaker.GetIndByName(b);
        if (o == 0) us.UnionV2(indexa, indexb);
        else {
            int roota = us.FindV2(indexa), rootb = us.FindV2(indexb);
            string ans = roota == rootb ? "yes" : "no";
            cout << ans << endl;
        }
    }
}


/**

10
0 Steven David
0 Lcch Dzx
1 Lcch Dzx
1 David Dzx
0 Lcch David
0 Frank Dzx
1 Steven Dzx
1 Frank David
0 Steven Dzx
0 Dzx Frank
样例输出
yes
no
yes
yes

*/
